### DISCUSSION TITLE
Get column (and table) names without access Database

### DISCUSSION THREAD

**Discussion by CharlotteJu** [simolus3/drift#2558] [Q&A] [answered]
Hello,

Is it possible to access column (and table) names outside of database object access?

I tried to directly call table().column.name:

Example:

```
class Users extends Table {
  IntColumn get id => integer()();
  TextColumn get lastName => text()();
}
```

In my database, I return a QueryRow (for my app's needs, as loading all objects would be too heavy).

In my widget, I wanted to do:

`int id = queryRow.read(Users().id.name);`

However, I encounter an error thrown by the dsl.dart file:
```
Never _isGenerated() {
  throw UnsupportedError(
    'This method should not be called at runtime. Are you sure you re-ran the '
    'builder after changing your tables or databases?',
  );
}
```
I tried to clean, invalidate the cache of Android Studio, and I have successfully generated the Drift file.

Note: I apologize for coming to ask a question again, but I have searched everywhere without success :(

**Comment by simolus3** [comment]
No worries! The idea is that the table classes you write are extended by drift to provide all the details about collected columns. So you can't construct an instance of your `Users` table directly.

> `int id = queryRow.read(Users().id.name);`

Unfortunately you need a reference to the database for that. If you have one, using `queryRow.read(database.users.id)` is no problem. But you'd have to pass a reference to the database around to the widget for that to work.

**Comment by CharlotteJu** [comment]
Thank you for your feedback.
Well, that won't work for me. 

So, I thought about creating static variables to be able to call the field name and make sure it's the right name : 

```
class Users extends Table {
 static const userId = 'id';
 static const userLastName = 'lastName';

  IntColumn get id => integer().named(userId)();
  TextColumn get lastName => text().named(userLastName)();
}
```

But I'm getting this error that's coming up...

`This table name is cannot be resolved! Please only use a constant string as parameter for .named().`

Is it not possible to use the named() method with a referenced string?

**Reply by simolus3** [reply]
>  Well, that won't work for me.

Could you describe why?

> Is it not possible to use the named() method with a referenced string?

No, it needs to be a direct string literal at the moment.

> I return a QueryRow (for my app's needs, as loading all objects would be too heavy)

Note that, as you get a `QueryRow` from drift, row objects for joined tables have already been loaded in there, so there is no strong performance benefit of passing a mostly unstructed `QueryRow` around. If you don't want to expose the database, you could export a record or a custom class for the select? E.g. if you have a `QueryRow` with two tables `a` and `b`, use a `(ARow, BRow)` type as a result instead?

If you have an example for the query you're running and what subsets of the data you don't want to parse unconditionally, I could look at better ways to approach them. At the moment I don't really understand your requirements.

**Reply by CharlotteJu** [reply]
Hello,

I'm sorry for the delay; we had an extended weekend (in France).

I use a QueryRow (occasionally) because I have large databases with many Foreign Keys, and returning complete objects causes significant lag. I also considered the idea of creating a Custom class to parse the result, but I have various queries depending on multiple parameters. That would require me to create a lot of custom classes for handling query results.

I cannot share my actual code here (as I work for a company), but if I provide an example of what I might have:

```
class User:
int id
String firstName
String lastName
int companyId
int professionId
int chiefId
int cityId
int projectId

class Company:
int id
String name
And 10 other attributes

class Profession:
int id
String name
And 10 other attributes

class City:
int id
String name
And 10 other attributes

class Project:
int id
String name
And 10 other attributes
```

And my query:
```
final uTable = database.users;
final coTable = database.companies;
final profTable = database.professions;
final ciTable = database.cities;
final projTable = database.projects;

final query = """
        SELECT  
        ${uTable.aliasedName}.${uTable.id.name} AS ${uTable.aliasedName}${uTable.id.name},
        ${uTable.aliasedName}.${uTable.firstName.name} AS ${uTable.aliasedName}${uTable.firstName.name},
        ${uTable.aliasedName}.${uTable.lastName.name} AS ${uTable.aliasedName}${uTable.lastName.name},
        ${coTable.aliasedName}.${coTable.name.name} AS ${coTable.aliasedName}${coTable.name.name},
        ${profTable.aliasedName}.${profTable.name.name} AS ${profTable.aliasedName}${profTable.name.name},
        ${ciTable.aliasedName}.${ciTable.name.name} AS ${ciTable.aliasedName}${ciTable.name.name},
        ${projTable.aliasedName}.${projTable.name.name} AS ${projTable.aliasedName}${projTable.name.name},
        /* AND OTHER ATTRIBUTES FROM TABLES */
        FROM ${uTable.aliasedName} 
        LEFT JOIN ${coTable.aliasedName} ON ${coTable.aliasedName}.${coTable.id.name} = ${uTable.aliasedName}.${uTable.companyId.name} 
        LEFT JOIN ${profTable.aliasedName} ON ${profTable.aliasedName}.${profTable.id.name} = ${uTable.aliasedName}.${uTable.professionId.name} 
        LEFT JOIN ${ciTable.aliasedName} ON ${ciTable.aliasedName}.${ciTable.id.name} = ${uTable.aliasedName}.${uTable.cityId.name} 
        LEFT JOIN ${projTable.aliasedName} ON ${projTable.aliasedName}.${projTable.id.name} = ${uTable.aliasedName}.${uTable.projectId.name} 
        /* WHERE MULTIPLE CONDITIONS */
        """;
```

And in my view, I wanted to do:
```
final uTable = Users()
Text(queryRow.read(${uTable.aliasedName}.${uTable.id.name} ));
/* AND OTHER LINES/*
```

**Reply by simolus3** [reply]
The easiest way to do this might be to return a record of the query row and the aliased tables that have been added to it, e.g.

```dart
typedef MyQueryResult = ({QueryRow row, TableInfo<Users, User> user, ...})
```

Then you could use `final user = data.user; data.queryRow.read('${uTable.aliasedName}.${uTable.id.name}')`. But I also want to mention that this is not an efficient way to use drift. Drift can construct the query with all relevant columns for you. You could for instance add a conditional `join()` based on flags passed to the method and only join the tables you really want.

**Reply by CharlotteJu** [reply]
Hello,

Thank you for the tip.

Regarding the Drift syntax, I have no issues using joins, but I encountered a problem when I wanted to return only certain columns.

For example, using the classes from my previous message:

I would like to return only the following fields:

class User: id, firstName, lastName
class Company: name
class Profession: name
class City: name
class Project: name

In Views, it's possible to specify the columns to return, but I haven't found a way to do it in a query when directly calling the database.

`select([uTable.id, uTable.firstName, uTable.lastName, coTable.name, ...]).from(uTable)...`

Is this possible?

**Reply by CharlotteJu** [reply]
Hello,

Thank you for the tip.

Regarding the Drift syntax, I have no issues using joins, but I encountered a problem when I wanted to return only certain columns.

For example, using the classes from my previous message:

I would like to return only the following fields:

class User: id, firstName, lastName
class Company: name
class Profession: name
class City: name
class Project: name

In Views, it's possible to specify the columns to return, but I haven't found a way to do it in a query when directly calling the database.

`select([uTable.id, uTable.firstName, uTable.lastName, coTable.name, ...]).from(uTable)...`

Is this possible?

**Reply by simolus3** [reply, marked as answer]
Yes! You can use the [`selectOnly`](https://drift.simonbinder.eu/api/drift/databaseconnectionuser/selectonly) method for that, e.g. with

```dart
final query = selectOnly(uTable)
    ..join(...)
    ..addColumns([uTable.id, uTable.firstName, uTable.lastName, ...]);
```
